bpf/bpf_host.c

```diff
@@ -800,9 +1061,47 @@ int to_host(struct __ctx_buff *ctx)
 	}
 
 	if (!traced)
-		send_trace_notify(ctx, TRACE_TO_STACK, src_label, 0, 0,
+		send_trace_notify(ctx, TRACE_TO_STACK, srcID, 0, 0,
 				  CILIUM_IFINDEX, ret, 0);
 
+#ifdef ENABLE_HOST_FIREWALL
+	if (!validate_ethertype(ctx, &proto)) {
+		ret = DROP_UNSUPPORTED_L2;
+		goto out;
+	}
+
+	policy_clear_mark(ctx);
+
+	switch (proto) {
+# if defined ENABLE_ARP_PASSTHROUGH || defined ENABLE_ARP_RESPONDER
+	case bpf_htons(ETH_P_ARP):
+		ret = CTX_ACT_OK;
+		break;
+# endif
+# ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		// TODO ipv6_host_policy_ingress.
+		ret = CTX_ACT_OK;
+		break;
+# endif
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		ret = ipv4_host_policy_ingress(ctx, &srcID);
+		break;
+# endif
```

Comment:
The "redirect to proxy" calls above do not steal the packet away to redirect to the proxy today, so I would expect this to break pod L7 policy if you enable host firewall policies. Presumably we should carefully detect that case and bypass the host firewall?

Reply:
If I understand correctly, these packets (with `MARK_MAGIC_TO_PROXY`) are on their way to the proxy from an endpoint, and you propose we whitelist that path (`ep -> proxy`)? Wouldn't we also need to whitelist the return path (`proxy -> ep`)?

I'm really unfamiliar with the L7 proxy stuff, but couldn't we express this whitelisting as a host policy (ingress to a specific port maybe)? Then we could have it as a sort of default host policy.

Reply:
> If I understand correctly, these packets (with MARK_MAGIC_TO_PROXY) are on their way to the proxy from an endpoint, and you propose we whitelist that path (ep -> proxy)? Wouldn't we also need to whitelist the return path (proxy -> ep)?

:+1: 

Forward path here should be straightforward. The knowledge we need is just a few lines above this point.

Traffic from the proxy we can identify using the packet mark (Proxy will always configure the SO_MARK on the socket; this is also how we inherit the original endpoint identity). Looking at the "inherit identity from stack" functions should provide you the context here.

> I'm really unfamiliar with the L7 proxy stuff, but couldn't we express this whitelisting as a host policy (ingress to a specific port maybe)? Then we could have it as a sort of default host policy.

Plausibly but a couple of caveats:
(1) The proxy port is not in the packet itself; TProxy allows us to redirect traffic without modifying the packet tuple.
(2) Adding the port to the proxy would presumably allow any peer to send traffic towards the proxy. It wouldn't just be traffic to a pod that is subject to L7 policy. So the proxy allow could potentially be broader than necessary.

Reply:
I pushed a fix for this in the new commit `bpf: Bypass host policies for packets to/from proxy`. I haven't validated it yet (besides `-C bpf build_all`), but pushed anyway to run CI tests. I'll validate manually in parallel.

Reply:
As discussed offline and commented by Thomas below, that commit to bypass the host policies for packets to/from proxy isn't needed:
- Packets redirected to the L7 proxy will have the IPs of the pods and therefore will already bypass the host policies.
- Packets from the proxy to the server (establishment of TCP handshake once handshake with client pod is done) will have the identity of the client pod embedded in their mark and that will be resolved correctly by the current bpf_host code, thereby bypassing the host policies again.
- Packets on the return path (e.g., TCP SYN+ACK from server pod to proxy) will match a CT entry inserted by the forward path packet.